### ISSUE TITLE
Consider send celery event using redis stream?

### ISSUE THREAD

**Issue by laixintao** [celery/celery#5317]
# Checklist
* [x] I have checked the issues list for similar or identical feature requests.
* [x] I have checked the commit log to find out if a feature was already implemented in master.

# Brief Summary

I setup my own celery `event` consumer (using redis as broker), but I can only have 1 process consume the event, if the process is down, then will cause outage. I can't set multiple event consumer because redis' pub/sub can not guarantee https://github.com/antirez/redis/issues/1974 multiple instance consume the event only once.

Can we provide a feature that celery worker(along with beat) can send event using redis stream? https://redis.io/topics/streams-intro

# Design
Provide a settings specific that send event to redis' pub/sub or stream for forward compatbility.

## Architectural Considerations

Is there a way that I can use redis as broker but use RabbitMQ sending event?

Any possible solution are appreciated.

**Comment by auvipy** [comment]
Do you have any implementation detail in mind? Could consider this for ceery 5.0

**Comment by laixintao** [comment]
Yes but not in detail, I want to reimplement the event handler and sender by redis stream, and give user an option to choose redis' stream or pub/sub as an broker for event.

I can give it a try.

**Comment by auvipy** [comment]
you can write a basic cep with basic proof of concept code example first in celery/cep repo

**Comment by jheld** [comment]
I'm super interested in this (ability to select, too).

I would prefer this go a step further (perhaps after it works): broker logic to use streams, too. I haven't checked the implementation so I don't know how similar events, pidbox, broker etc is. Though I think the pidbox/reply queue would be the same as normal broker?

**Comment by jheld** [comment]
Just an update here: still interested. Depending on how active my other responsibilities at work are (barring free time off hours), I will look into this design & code wise.

Seems using streams would be compatible (better?) in many things, including how redis visibility_timeout flow operates, and acking in general. One issue is around the notion of dead consumers, but that was always a problem with redis, so if streams improves enough, then allowing that usage is probably for the best, and all clients should eventually be using redis 5+ (deprecation timeline _will_ happen).

**Comment by laixintao** [comment]
@jheld Yes, stream seems perfect for a broker from my point of view.

About the dead consumers. I think we can let other workers check the streams using `xpending` command, and then `xclaim` those pending messages.

Stream's group/ack/xrange query based on time looks awesome, I'd like to discuss more with those ideas. :)

Also, I wrote a better Redis client: https://github.com/laixintao/iredis  You may like this if you debug stream a lot.

**Comment by jheld** [comment]
Right. I just meant that determining _when_ do XCLAIM will likely still be based on something like visibility_timeout etc.

So implementation wise trying it on events first seems like a good grass roots experiment. I can write up a CEP against the code change and see what happens from there.

**Comment by jheld** [comment]
Partial update: I have made bare-bones on the fly ability to use streams (along with pub/sub) for events dispatcher and events receiver. needs to be coded out more explicitly, but wanted to see what would happen.

Redis Pub/Sub does not appear to store messages anywhere -- just sent down a socket basically.

So if flower for instance is not connected, then the messages just die/aren't sent at all.

Streams will allow us to hold those messages around in case flower/consumer isn't around/has a bug/crash.

But, that means that the application/flower will have to determine a proper xdel/xtrim/xadd (deletion after a certain stream size) policy.

We don't want redis streams to topple the RAM of redis. Due to that issue, there would be a noted documentation caveat regarding examples/expectations regarding how to trim/remove items from a given stream. This is definitely on the ops/infrastructure side of things but it's something that must be discussed at least somewhere.

We _could_ provide a simple interface which allows the application/celery setting to determine an "appropriate" stream size in which xadd will just start trimming as it adds (part of redis' flow), and probably a default of it not being active. I'd rather all data in the general case vs people not being aware that we're deleting data behind the scenes. Again, documentation.

**Comment by jheld** [comment]
Reading through your description, there was an ask to allow pub/sub to try to do single consumer, but pub/sub is written for broadcast (I alluded to in my above comment).

You should be able to have multiple flower running at the same time and that _may_ help reduce downtime/outage. But yes, it's not perfect, and streams appears to be a better solution so far.

**Comment by gee-gendo** [comment]
Hi, any plans for this? Is there another more recent ticket related to this?

**Comment by auvipy** [comment]
it would be great if we can get a draft PR
